Core/Core/Files/FileDetails/FileDetailsViewController.swift

```diff
@@ -0,0 +1,365 @@
+            guard webView.estimatedProgress >= 1 else { return }
+            self?.loadObservation = nil
+            self?.doneLoading()
+        }
+        webView.loadFileURL(url, allowingReadAccessTo: url)
+    }
+
+    func doneLoading() {
+        activityView.stopAnimating()
+        progressView.isHidden = true
+        NotificationCenter.default.post(name: .init("CBIModuleItemProgressUpdatedNotification"), object: nil, userInfo: [
+            "CBIUpdatedModuleItemIDStringKey": fileID,
+            "CBIUpdatedModuleItemTypeKey": "must_view",
+        ])
+    }
+
+    @IBAction func viewModules() {
+        env.router.route(to: Route.modules(forCourse: context.id), from: self, options: nil)
+    }
+
+    @objc func share(_ sender: UIBarButtonItem) {
+        guard let url = localURL else { return }
+        let controller = UIActivityViewController(activityItems: [url], applicationActivities: nil)
+        controller.popoverPresentationController?.barButtonItem = sender
+        env.router.show(controller, from: self, options: .modal)
+    }
+}
+
+extension FileDetailsViewController: URLSessionDownloadDelegate {
+    var localURL: URL? {
+        guard let sessionID = env.currentSession?.uniqueID, let name = files.first?.filename else { return nil }
+        let base = files.first?.mimeClass == "pdf" ? URL.documentsDirectory : URL.temporaryDirectory
+        return base.appendingPathComponent("\(sessionID)/\(fileID)/\(name)")
+    }
+
+    func downloadFile(at url: URL) {
+        if let path = localURL?.path, FileManager.default.fileExists(atPath: path) { return downloadComplete() }
+        downloadTask = URLSessionAPI.delegateURLSession(.ephemeral, self, nil).downloadTask(with: url)
+        downloadTask?.resume()
+    }
+
+    public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        performUIUpdate {
+            self.progressView.setProgress(Float(totalBytesWritten) / Float(totalBytesExpectedToWrite), animated: true)
+        }
+    }
+
+    public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
+        guard let localURL = localURL else { return }
+        do {
+            try FileManager.default.createDirectory(at: localURL.deletingLastPathComponent(), withIntermediateDirectories: true)
+            if FileManager.default.fileExists(atPath: localURL.path) {
+                try FileManager.default.removeItem(at: localURL)
+            }
+            try FileManager.default.moveItem(at: location, to: localURL)
+        } catch {
+            showError(error)
+        }
+    }
+
+    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        if let error = error, (error as NSError).code != -999 { showError(error) }
```

Comment:
What's the -999 about? quick google search didn't tell me what that code is.

Reply:
Oh I see from the test now that it is cancel.

Reply:
would a comment or a named constant be helpful here?

Reply:
Up to you. This is fine with me.